test/JoinTest.cpp

```diff
@@ -0,0 +1,429 @@
+  // For generating better messages, when failing a test.
+  auto trace{generateLocationTrace(l, "runTestCasesForAllJoinAlgorithm")};
+ 
+  // All normal join algorithm defined as lambda functions for easier handing
+  // over to helper functions.
+  Join J{Join::InvalidOnlyForTestingJoinTag{}};
+  auto HashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.hashJoin<A, B, C>(AD_FWD(args)...);
+  };
+  auto JoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+
+  // For sorting IdTables by their join column.
+  auto sortByJoinColumn = [](IdTable& idTable, size_t joinColumn) {
+    /*
+     * TODO Introduce functionality to the IdTable-class, so that std::ranges::sort
+     * can be used instead of std::sort. Currently it seems like the iterators
+     * , produced by IdTable, aren't the right type.
+     */
+    std::sort(idTable.begin(), idTable.end(), [&joinColumn](const auto& row1,
+          const auto& row2) {return row1[joinColumn] < row2[joinColumn];});
+  };
+
+  // Random shuffle both tables, run hashJoin, check result.
+  for (normalJoinTest& testCase: testSet) {
+    randomShuffle(testCase.leftInput.idTable.begin(), testCase.leftInput.idTable.end());
+    randomShuffle(testCase.rightInput.idTable.begin(), testCase.rightInput.idTable.end());
+    testCase.resultMustBeSortedByJoinColumn = false;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
+
+  // Sort the larger table by join column, run hashJoin, check result (this time it's sorted).
+  for (normalJoinTest& testCase: testSet) {
+    if (testCase.leftInput.idTable.size() >=
+        testCase.rightInput.idTable.size()) {
+      sortByJoinColumn(testCase.leftInput.idTable, testCase.leftInput.joinColumn);
+    } else {
+      sortByJoinColumn(testCase.rightInput.idTable, testCase.rightInput.joinColumn);
+    }
+
+    testCase.resultMustBeSortedByJoinColumn = true;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
+
+  // Sort both tables, run merge join and hash join, check result. (Which has to be sorted.)
+  for (normalJoinTest& testCase: testSet) {
+    sortByJoinColumn(testCase.leftInput.idTable, testCase.leftInput.joinColumn);
+    sortByJoinColumn(testCase.rightInput.idTable, testCase.rightInput.joinColumn);
+    testCase.resultMustBeSortedByJoinColumn = true;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, JoinLambda);
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
+}
+
+/* 
+ * @brief Return a vector of normalJoinTest for testing with the join functions. 
+ */
+std::vector<normalJoinTest> createNormalJoinTestSet() {
+  std::vector<normalJoinTest> myTestSet{};
+
+  // For easier creation of IdTables and readability.
+  vectorTable leftIdTable{{{1, 1}, {1, 3}, {2, 1}, {2, 2}, {4, 1}}};
+  vectorTable rightIdTable{{{1, 3}, {1, 8}, {3, 1}, {4, 2}}};
+  vectorTable sampleSolution{{{1, 1, 3}, {1, 1, 8}, {1, 3, 3}, {1, 3, 8}, {4, 1, 2}}};
```

Comment:
not `sampleSolution` but `expectedResult`.

Reply:
Done, see [20b967e](https://github.com/ad-freiburg/qlever/pull/821/commits/20b967e9d00fa0bd30ac458253b949da2e2446ad).